### PR TITLE
http: add response flag on downstream max stream duration

### PR DESCRIPTION
**Pull request by Shikugawa**
Signed-off-by: Shikugawa <rei@tetrate.io>

For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/master/PULL_REQUESTS.md)

Commit Message: This patch is to resolve https://github.com/envoyproxy/envoy/issues/11706. To support returning response flags when downstream max stream duration reached.
Additional Description: In this PR, envoy will return 408 or arbitrary http status code when downstream response is started when downstream max_stream_duration timer will be invoked.
Risk Level: Low
Testing: Unit / Integration
Docs Changes:
Release Notes:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Deprecated:]

### CODE REVIEW

**Comment by Shikugawa** [comment]
> Thanks! Generally this looks great to me. Would you be able to add an end to end test testing if the response had started?

@asraa Done.

**Comment by Shikugawa** [comment]
@asraa ping?

**Comment by asraa** [comment]
@envoyproxy/senior-maintainers Could someone PTAL for merging? Thank you!

File: source/common/http/conn_manager_impl.cc [hunk 1]
```diff
@@ -679,7 +680,16 @@ void ConnectionManagerImpl::ActiveStream::onRequestTimeout() {
 void ConnectionManagerImpl::ActiveStream::onStreamMaxDurationReached() {
   ENVOY_STREAM_LOG(debug, "Stream max duration time reached", *this);
   connection_manager_.stats_.named_.downstream_rq_max_duration_reached_.inc();
-  connection_manager_.doEndStream(*this);
+  stream_info_.setResponseFlag(StreamInfo::ResponseFlag::DownstreamMaxStreamDurationReached);
+
+  // If downstream response had started, envoy don't return 408.
```
Comment (mattklein123) [inline source/common/http/conn_manager_impl.cc:685]: This is a behavior change  / bug fix and not just a response flag change. Please add a release note and cleanup the PR description to fully follow the template.

File: source/common/stream_info/utility.cc [hunk 2]
```diff
@@ -39,7 +39,7 @@ void ResponseFlagUtils::appendString(std::string& result, const std::string& app
 const std::string ResponseFlagUtils::toShortString(const StreamInfo& stream_info) {
   std::string result;
 
-  static_assert(ResponseFlag::LastFlag == 0x100000, "A flag has been added. Fix this code.");
+  static_assert(ResponseFlag::LastFlag == 0x200000, "A flag has been added. Fix this code.");
```
Comment (mattklein123) [inline source/common/stream_info/utility.cc:42]: Fix the code per the comment and add tests. Same elsewhere.

Comment (Shikugawa) [reply source/common/stream_info/utility.cc:42]: All of this type of code had already changed from `0x100000` to `0x200000`.

File: source/common/http/conn_manager_impl.cc (hunk 1 above)
Comment (Shikugawa) [reply source/common/http/conn_manager_impl.cc:685]: add release note.

Comment (alyssawilk) [reply source/common/http/conn_manager_impl.cc:693]: Heh, just discovered this myself and have a PR which also fixes, if you want to just handle the response flag here.
If you want to fix the local reply, you'll need to flag protect (As Matt said) and also please only use sendLocalReply - it handles the "reset or reply" logic internally now.

Comment (Shikugawa) [reply source/common/http/conn_manager_impl.cc:693]: Does it mean that we don't have to aware the downstream response_header status and just to call sendLocalReply?

Comment (mattklein123) [reply source/common/http/conn_manager_impl.cc:693]: Correct, I think you can now revert this part of the change but I would confirm with @alyssawilk

File: source/common/stream_info/utility.cc (hunk 2 above)
Comment (mattklein123) [reply source/common/stream_info/utility.cc:42]: Please read the files below and you will see there is more work to do in each of these places. If further comments would have helped you see this better please add them. Thank you!

File: source/common/http/conn_manager_impl.cc (hunk 1 above)
Comment (alyssawilk) [reply source/common/http/conn_manager_impl.cc:693]: So you have two choices.  You can either revert this part of the change (set response flags, and let my WIP PR take care of the sendLocalResponse changes) or you can runtime-guard always calling sendLocalReply as I do there.  calling doEndStream is basically deprecated.

Comment (alyssawilk) [reply source/common/http/conn_manager_impl.cc:693]: https://github.com/envoyproxy/envoy/pull/12228 sent out before I got all the test cleaned up, but it has the sendLocalReply with guards I mentioned.

**Pull request closed without merging**